Bài 7. File dữ liệu HocSinh.csv lưu thông tin của một nhóm học sinh về giới tính
(GioiTinh), lớp (Lop), tuổi (Tuoi), nơi ở (NoiO), học lực (HocLuc), mức độ yêu
thích thể thao (TheThao) và đánh giá hình thức (HinhThuc). Hãy lấy file dữ liệu
và thực hiện các yêu cầu sau:

## 1. Tính số học sinh nữ, nam.

Muốn tính số học sinh nữ, nam ta sẽ lập bảng tần số cho biến Giới tính

In [ ]:
# import dữ liệu từ file csv: Dùng method pandas.read_csv()
import pandas as pd
sv=pd.read_csv('HocSinh.csv')
print(sv)

FileNotFoundError: ignored

In [ ]:
# Lập bảng tần số cho biến giới tính
print(sv['GioiTinh'].value_counts())

Nu     248
Nam    227
Name: GioiTinh, dtype: int64


Vậy có 248 nữ và 227 nam.

In [ ]:
# Hiển thị các tham số của pd.value_counts
pd.value_counts

<function pandas.core.algorithms.value_counts(values, sort: 'bool' = True, ascending: 'bool' = False, normalize: 'bool' = False, bins=None, dropna: 'bool' = True) -> 'Series'>

In [ ]:
# Đọc chi tiết cách dùng pd.value_counts
?pd.value_counts

Object `pd.value_counts.cumsum` not found.


## 2. Tính số học sinh nam có mức độ yêu thích thể thao là 4.

Ta tìm kết quả bằng cách lập bảng tần số chéo của 2 biến là giới tính và thể thao.

In [ ]:
pd.crosstab(sv.GioiTinh,sv.TheThao)

TheThao     1   2   3   4
GioiTinh                 
Nam       127  60  31   9
Nu         38  93  80  37

Vậy có 9 học sinh nam có mức độ yêu thích thể thao là 4.

## 3. Tính tỷ lệ học sinh có mức độ yêu thích thể thao là 4.

Để tính tỷ lệ này ta lập tần suất cho biến thể thao

In [ ]:
sv.TheThao.value_counts(normalize = True)
# đặt normalize = True để tính tần suất

1    0.347368
2    0.322105
3    0.233684
4    0.096842
Name: TheThao, dtype: float64

Vậy tỉ lệ cần tìm là 0.096842 = 9.6842%

#### 3.1 Tính tỷ lệ học sinh từ 9 tuổi trở lên

In [ ]:
tuoi9 = sv.Tuoi >= 9
tuoi9.value_counts(normalize=True)

## 4. Tính tỷ lệ học sinh có mức độ yêu thích thể thao là 4 trong nhóm học sinh nam, trong nhóm học sinh nữ.

Để tìm các tỷ lệ này ta tính tần suất của bảng tần số chéo của 2 biến giới tính và thể thao. Đây là bảng tần số chéo đó.

In [ ]:
pd.crosstab(sv.TheThao,sv.GioiTinh,margins=True)
# đặt margins=True để tính tổng cột và tổng hàng

GioiTinh  Nam   Nu  All
TheThao                
1         127   38  165
2          60   93  153
3          31   80  111
4           9   37   46
All       227  248  475

Có 3 cách tính tần suất cho bảng này, được quy định bởi tham số normalize

 1. Tính tần suất đối với toàn bộ tập dữ liệu: normalize = 'all'
        
 2. Tính tần suất theo dòng: normalize = 'index'
    
 3. Tính tần suất theo cột: normalize = 'columns'

 Để tính tỷ lệ học sinh có mức độ yêu thích thể thao là 4 trong nhóm
học sinh nam, trong nhóm học sinh nữ ta tính tần suất theo cột, vì nam và nữ là 2 cột trong bảng tần số chéo trên.

In [ ]:
pd.crosstab(sv.TheThao,sv.GioiTinh,normalize='columns')

GioiTinh       Nam        Nu
TheThao                     
1         0.559471  0.153226
2         0.264317  0.375000
3         0.136564  0.322581
4         0.039648  0.149194

Tỷ lệ học sinh có mức độ yêu thích thể thao là 4 trong nhóm
học sinh nam, trong nhóm học sinh nữ lần lượt là 0.039648 và 0.149194

#### 4.1 Tính tỷ lệ học sinh từ 9 tuổi trở lên trong nhóm nam

In [ ]:
pd.crosstab(tuoi9,sv.GioiTinh,normalize='columns') # tuoi9 đã tạo ra trong mục 3.1

## 5. Tính tỉ lệ những học sinh là nam và có mức độ yêu thích thể thao bằng 4

In [ ]:
pd.crosstab(sv.TheThao,sv.GioiTinh,normalize='all')

GioiTinh       Nam        Nu
TheThao                     
1         0.267368  0.080000
2         0.126316  0.195789
3         0.065263  0.168421
4         0.018947  0.077895

Vậy tỉ lệ những học sinh là nam và có mức độ yêu thích thể thao bằng 4 là 0.018947

Bài 8. Trong file dữ liệu có tên là SoLieu.csv chứa một số thông tin cá nhân của
100 người về giới tính (GioiTinh), tuổi (Tuoi), khu vực sống (KhuVuc) và tổng thu
nhập (đơn vị triệu VND) trong năm 2008 (ThuNhap). Hãy lấy file dữ liệu và thực
hiện các yêu cầu sau:

1. Tiến hành phân tổ cột dữ liệu về tuổi thành các tổ với các điểm chia là 20,30,40,50,60,70,80
và tính tỉ lệ những người được điều tra có độ tuổi không vượt quá 50.

In [ ]:
sl=pd.read_csv('SoLieu.csv')

Các bước Phân tổ dữ liệu:
    
1. Tìm min, max

2. Xác định các điểm chia (Chú ý: điểm chia nhỏ nhất <= min <= max <= điểm chia lớn nhât)

3. Dùng lệnh pd.cut để phân tổ dữ liệu

In [ ]:
min(sl.Tuoi)

24

In [ ]:
max(sl.Tuoi)

78

In [ ]:
?pd.cut

In [ ]:
ptTuoi=pd.cut(sl.Tuoi,bins=[20, 30, 40, 50, 60, 70, 80],
              include_lowest=True)
# bins: các điểm chia
# include_lowest = True: làm cận dưới của tổ đầu tiên giảm đi
# một ít (để chứa cả giá trị 20)

In [ ]:
# tính tỉ lệ những người được điều tra có độ tuổi không vượt quá 50.
ptTuoi.value_counts(sort = False, normalize = True).cumsum()
# sort = False để xếp các tổ theo chiều tăng của dữ liệu,
# ngược lại theo chiều giảm của tần số (tần suất)
# cumsum để tính tần suất tích lũy

(19.999, 30.0]    0.20
(30.0, 40.0]      0.61
(40.0, 50.0]      0.73
(50.0, 60.0]      0.85
(60.0, 70.0]      0.98
(70.0, 80.0]      1.00
Name: Tuoi, dtype: float64

Vậy tỉ lệ này là 73%

2. Tiến hành phân tổ cột dữ liệu về thu nhập thành các tổ với các điểm chia là
20,40,60,80,100

In [ ]:
min(sl.ThuNhap)

20

In [ ]:
max(sl.ThuNhap)

100

In [ ]:
ptTN=pd.cut(sl['ThuNhap'],bins=[20,40,60,80,100],include_lowest='True')

3. Tính tỉ lệ những người phải đóng thuế thu nhập nếu biết một người phải đóng
thuế thu nhập nếu tổng thu nhập trong năm của người đó vượt quá 60
triệu VND.

In [ ]:
ptTN.value_counts(normalize=True, sort = False).cumsum()

(19.999, 40.0]    0.47
(40.0, 60.0]      0.79
(60.0, 80.0]      0.93
(80.0, 100.0]     1.00
Name: ThuNhap, dtype: float64

Tỉ lệ là: 1 - 0.79 = 0.21 (Lấy 1 trừ đi tỉ lệ <= 60 thì ra tỉ lệ > 60)

4. Tính tỉ lệ những người có thu nhập hơn 80 triệu nằm trong khoảng tuổi (40,50].

In [ ]:
pd.crosstab(ptTuoi,ptTN,normalize='index')

ThuNhap         (19.999, 40.0]  (40.0, 60.0]  (60.0, 80.0]  (80.0, 100.0]
Tuoi                                                                     
(19.999, 30.0]        0.500000      0.250000      0.200000       0.050000
(30.0, 40.0]          0.463415      0.341463      0.121951       0.073171
(40.0, 50.0]          0.250000      0.416667      0.250000       0.083333
(50.0, 60.0]          0.583333      0.333333      0.083333       0.000000
(60.0, 70.0]          0.538462      0.230769      0.076923       0.153846
(70.0, 80.0]          0.500000      0.500000      0.000000       0.000000

Tỉ lệ cần tính là 0.083